### PR TITLE
feat(real-ip): support search recursive

### PR DESCRIPTION
**Pull request by crazyMonkey1995**
### Description

Support real ip recusive, see http://nginx.org/en/docs/http/ngx_http_realip_module.html

Fixes https://github.com/apache/apisix/issues/6602
### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/plugins/real-ip.lua [hunk 1]
```diff
@@ -67,6 +73,19 @@ function _M.check_schema(conf)
...
 
 
+local function addr_match(conf, addr)
+    if not conf.trusted_addresses then
+        return false
+    end
+
+    if not conf.matcher then
+        conf.matcher = core.ip.create_ip_matcher(conf.trusted_addresses)
+    end
```
Comment (tzssangglass) [inline apisix/plugins/real-ip.lua:83]: It is bad practice to add properties to conf, and we should keep conf pure. How about making `matcher` a module-level variable?

Comment (crazyMonkey1995) [reply apisix/plugins/real-ip.lua]: Changed it to local variable.

File: apisix/plugins/real-ip.lua [hunk 2]
```diff
@@ -84,6 +104,22 @@ local function get_addr(conf, ctx)
...
         end
 
+        if conf.recursive and conf.trusted_addresses then
+            local split_addrs = {}
+            local match_itr = str_gmatch(addrs, "[^,%s*]+")
+            for itr in match_itr do
+                tb_insert(split_addrs, itr)
+            end
+
+            for i = #split_addrs, 1, -1 do
```
Comment (spacewander) [inline apisix/plugins/real-ip.lua:114]:
```suggestion
            for i = #split_addrs, 2, -1 do
```
We don't care if the first one can be trusted.

File: apisix/plugins/real-ip.lua [hunk 3]
```diff
@@ -84,6 +104,22 @@ local function get_addr(conf, ctx)
             return addrs
         end
 
+        if conf.recursive and conf.trusted_addresses then
+            local split_addrs = {}
+            local match_itr = str_gmatch(addrs, "[^,%s*]+")
```
Comment (tzssangglass) [inline apisix/plugins/real-ip.lua:109]: can we use `ngx.re.gmatch` to replace `string.gmatch`

**Comment by tzssangglass** [comment]
LGTM

**Review by spacewander — DISMISSED** [review]

File: apisix/plugins/real-ip.lua (hunk 3 above)
Comment (crazyMonkey1995) [reply apisix/plugins/real-ip.lua]: Solved.

File: apisix/plugins/real-ip.lua [hunk 4]
```diff
@@ -106,7 +106,7 @@ local function get_addr(conf, ctx)
 
         if conf.recursive and conf.trusted_addresses then
             local split_addrs = {}
-            local match_itr = str_gmatch(addrs, "[^,%s*]+")
+            local match_itr = ngx.re.gmatch(addrs, "[^,%s*]+", "jo")
```
[the remark this answers is not in the document]
Comment (spacewander) [reply apisix/plugins/real-ip.lua:109]: And the return value is different.

[the remark this answers is not in the document]
Comment (crazyMonkey1995) [reply apisix/plugins/real-ip.lua:109]: Sorry for that, I should have tested it first

[the remark this answers is not in the document]
Comment (crazyMonkey1995) [reply apisix/plugins/real-ip.lua]: I replaced it with ngx.re.split

File: apisix/plugins/real-ip.lua [hunk 5]
```diff
@@ -84,6 +103,17 @@ local function get_addr(conf, ctx)
             return addrs
         end
 
+        if conf.recursive and conf.trusted_addresses then
+            local split_addrs, _ = ngx_re_split(addrs, ",\\s*", "jo")
```
Comment (soulbird) [inline apisix/plugins/real-ip.lua:107]: `local split_addrs = ngx_re_split(addrs, ",\\s*", "jo")` better? Or deal with the error?

File: apisix/plugins/real-ip.lua [hunk 6]
```diff
@@ -105,12 +135,8 @@ function _M.rewrite(conf, ctx)
...
-        if not conf.matcher then
-            conf.matcher = core.ip.create_ip_matcher(conf.trusted_addresses)
-        end
-
         local remote_addr = ctx.var.remote_addr
-        local trusted = conf.matcher:match(remote_addr)
+        local trusted = addr_match(conf, remote_addr)
         if not trusted then
             return
         end
```
Comment (soulbird) [inline apisix/plugins/real-ip.lua:142]:
```lua
if not addr_match(conf, remote_addr) then
    return
end
```

File: apisix/plugins/real-ip.lua (hunk 5 above)
Comment (crazyMonkey1995) [reply apisix/plugins/real-ip.lua:107]: Thanks for the review, the above if condition has already dealt with the absence of commas, so there should be no need to deal with err here? I'm not very familiar with lua, but this `_` may indeed be removed.

**Review by tzssangglass — APPROVED** [review]

File: apisix/plugins/real-ip.lua [hunk 7]
```diff
@@ -67,6 +73,19 @@ function _M.check_schema(conf)
 end
 
 
+local function addr_match(conf, addr)
+    if not conf.trusted_addresses then
+        return false
+    end
+
+    if not matcher then
```
Comment (spacewander) [inline apisix/plugins/real-ip.lua:81]: We should not use global matcher. Otherwise different configuration will use the same matcher.

Comment (crazyMonkey1995) [reply apisix/plugins/real-ip.lua:81]: But I defined it as a local variable above.

Comment (spacewander) [reply apisix/plugins/real-ip.lua:81]: Different requests with different configurations will share this as it is defined as a module variable.

Comment (crazyMonkey1995) [reply apisix/plugins/real-ip.lua:81]: I see, can I use lru cache to solve it? Or do you have any suggestions?

Comment (crazyMonkey1995) [reply apisix/plugins/real-ip.lua:85]: Use lru cache instead, refer to the implementation of ip-restriction

File: apisix/plugins/real-ip.lua [hunk 8]
```diff
@@ -67,6 +75,22 @@ function _M.check_schema(conf)
 end
 
 
+local function addr_match(conf, addr)
+    if not conf.trusted_addresses then
+        return false
+    end
+
+    local matcher, err = lrucache(conf.trusted_addresses, nil,
+                                  core.ip.create_ip_matcher, conf.trusted_addresses)
```
[the remark this answers is not in the document]
Comment (spacewander) [reply apisix/plugins/real-ip.lua:84]: Better to use https://github.com/apache/apisix/blob/4690feb421779f5b79e8dd990dc00f4d3f1052d0/apisix/core/lrucache.lua#L171 instead.

File: apisix/plugins/real-ip.lua [hunk 9]
```diff
@@ -67,6 +75,22 @@ function _M.check_schema(conf)
 end
 
 
+local function addr_match(conf, ctx, addr)
+    if not conf.trusted_addresses then
```
Comment (spacewander) [inline apisix/plugins/real-ip.lua:79]: We already check the trusted_addresses before: https://github.com/apache/apisix/blob/025710564ac717f99fb96dd19dc1fdd1344bc7ef/apisix/plugins/real-ip.lua#L107?

**Review by tzssangglass — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**